### PR TITLE
Enhance shadowsocks.init robustness

### PR DESCRIPTION
**Pull request by MartinWetterwald**
# TODO:

* ~~Bug when calling reload if shadowsocks is stopped~~
* ~~Need to update hotplug script referencing old name -j SOCKS~~
* ~~Need to update /sbin/overthebox_test_download_proof~~
* ~~Make ifup idempotent~~
* ~~Find a way to count CPU cores and not CPU threads~~
* ~~Validate uci config before doing anything~~
* ~~Add EXTRA_HELP~~
* ~~Embed a documentation of the used iptables/ipset layout~~
* ~~Correct spelling mistakes~~
* ~~Remove obsolete tracker parameters (reliability, up, down, count)~~
* ~~Use "." command instead of "source" for compat~~
* ~~Put each member of LOCALNETS on a separated line~~
* ~~Bug when calling ifup/ifdown if shadowsocks is stopped~~

### CODE REVIEW

File: shadowsocks-libev/files/shadowsocks.init [hunk 1]
```diff
@@ -1,298 +1,512 @@
...
+_launch_shadow() {
+    local conf cpu_count
+
+    # Determine the number of CPU cores
+    local cpu_sysfs="/sys/devices/system/cpu/"
+
+    if [ -d $cpu_sysfs ]; then
+        # Try to get the number of CPU cores
+        # We need to add a backslash before the + because busybox's find uses "posix-basic" regextype
+        cpu_count=$(find $cpu_sysfs -maxdepth 1 -type d -regex '.*/cpu[0-9]\+' | wc -l)
```
Comment (sduponch) [inline shadowsocks-libev/files/shadowsocks.init]: This count dies cores and threads cores, not only real cores : You must count uniq cores id per physicals id.

File: shadowsocks-libev/files/shadowsocks.init [hunk 2]
```diff
@@ -1,298 +1,512 @@
...
-ifdown() {
-	shadowsocks_disable_interfaces_hook "$@"
+# $1 is a ref to the current zone section being processed
+_update_or_enable_hook_one_zone() {
+    # Let's fetch the current zone name
+    local zone
+    config_get zone $1 name
+
+    # Only traffic coming from a lan zone's interface will be routed to shadowsocks
+    if [ "$zone" = "lan" ]; then
```
Comment (sduponch) [inline shadowsocks-libev/files/shadowsocks.init]: i sugesst to add an uci shadowsock list entry to add more zone than "lan", where value the lan is is the default.

File: shadowsocks-libev/files/shadowsocks.init (hunk 1 above)
Comment (MartinWetterwald) [inline shadowsocks-libev/files/shadowsocks.init]: Done. I replaced "sort | uniq" by "sort -u".

File: shadowsocks-libev/files/shadowsocks.init (hunk 2 above)
Comment (MartinWetterwald) [inline shadowsocks-libev/files/shadowsocks.init]: Maybe for a next release :)

**Pull request merged**